src/execution_plan/optimizations/apply_join.c

```diff
@@ -0,0 +1,162 @@
+static bool _applicableFilter(const FT_FilterNode *f) {
+    return (f->t == FT_N_PRED && f->pred.op == EQ);
+}
+
+// Collects all consecutive filters beneath given op.
+static Filter** _locate_filters(OpBase *cp) {
+    OpBase *parent = cp->parent;
+    Filter **filters = array_new(Filter*, 0);
+
+    while(parent && parent->type == OPType_FILTER) {
+        filters = array_append(filters, (Filter*)parent);
+        parent = parent->parent;
+    }
+
+    return filters;
+}
+
+// Tests if stream resolves all aliases.
+static bool _stream_resolves_entities(const OpBase *root, rax *aliases) {
+    if(root->modifies) {
+        for(int i = 0; i < Vector_Size(root->modifies); i++) {
+            char *alias;
+            Vector_Get(root->modifies, i, &alias);            
+            raxRemove(aliases, (unsigned char*)alias, strlen(alias), NULL);
+        }
+    }
+
+    if(raxSize(aliases) == 0) return true;
+
+    for(int i = 0; i < root->childCount; i++) {
+        if(_stream_resolves_entities(root->children[i], aliases)) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+/* filter is composed of two expressions:
+ * left and right hand side
+ * to apply a join operation each expression
+ * must be entirely resolved by one of the branches
+ * either left or right of the cartesian product operation
+ * _relate_exp_to_stream will try to associate each expression
+ * lhs and rhs with the appropriate branch. */
+static void _relate_exp_to_stream(const OpBase *cp, const FT_FilterNode *f, AR_ExpNode **lhs, AR_ExpNode **rhs) {
+    *lhs = NULL;
+    *rhs = NULL;
+    bool expression_resolved = true;
+
+    AR_ExpNode *lhs_exp = f->pred.lhs;
+    AR_ExpNode *rhs_exp = f->pred.rhs;
+
+    assert(cp->childCount == 2);
+    OpBase *left_child = cp->children[0];
+    OpBase *right_child = cp->children[1];
+
+    rax *aliases = raxNew();
+
+    /* Make sure LHS and RHS expressions can be resolved.
+     * Left expression - Left branch
+     * Right expression - Right branch */
+    AR_EXP_CollectAliases(lhs_exp, aliases);    
+    if(_stream_resolves_entities(left_child, aliases)) {
+        // aliases is now empty.
+        AR_EXP_CollectAliases(rhs_exp, aliases);        
+        if(_stream_resolves_entities(right_child, aliases)) {            
+            *lhs = lhs_exp;
+            *rhs = rhs_exp;
+            raxFree(aliases);
+            return;
+        }
+    }
+
+    /* Left expression - Right branch
+     * Right expression - Left branch */
+    *lhs = NULL;
+    *rhs = NULL;
+
+    raxFree(aliases);
+    aliases = raxNew();
+
+    AR_EXP_CollectAliases(lhs_exp, aliases);
+    if(_stream_resolves_entities(right_child, aliases)) {
+        // aliases is now empty.
+        AR_EXP_CollectAliases(rhs_exp, aliases);
+        if(_stream_resolves_entities(left_child, aliases)) {
+            *lhs = rhs_exp;
+            *rhs = lhs_exp;
+            raxFree(aliases);
+            return;
+        }
+    }
```

Comment:
I think this function would be more readable if we could make 2 separate calls, first trying to resolve all expressions on left/left, then on left/right if that fails. It's not a significant issue, though.